spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningMetadataSuite.scala

```diff
@@ -0,0 +1,505 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta
+
+import org.apache.spark.sql.delta.test.DeltaSQLCommandTest
+
+import org.apache.spark.sql.QueryTest
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.types._
+
+/**
+ * Suite covering the[[TypeWideningMetadata]] and [[TypeChange]] classes used to handle the metadata
+ * recorded by the Type Widening table feature in the table schema.
+ */
+class DeltaTypeWideningMetadataSuite extends QueryTest with DeltaSQLCommandTest {
+  private val testTableName: String = "delta_type_widening_metadata_test"
+
+  /** A dummy transaction to be used by tests covering `addTypeWideningMetadata`. */
+  private lazy val txn: OptimisticTransaction = {
+    val table = spark.sessionState.catalog.getTableMetadata(TableIdentifier(testTableName))
+    DeltaLog.forTable(spark, TableIdentifier(testTableName))
+        .startTransaction(catalogTableOpt = Some(table))
+  }
+
+  override protected def beforeAll(): Unit = {
+    super.beforeAll()
+    sql(s"CREATE TABLE $testTableName (a int) USING delta " +
+      s"TBLPROPERTIES ('${DeltaConfigs.ENABLE_TYPE_WIDENING.key}' = 'true')")
+  }
+
+  override protected def afterAll(): Unit = {
+    try {
+      sql(s"DROP TABLE IF EXISTS $testTableName")
+    } finally {
+      super.afterAll()
+    }
+  }
+
+  /**
+   * Short-hand to build the metadata for a type change to cut down on repetition.
+   */
+  private def typeChangeMetadata(
+      version: Long,
+      fromType: String,
+      toType: String,
+      path: String = ""): Metadata = {
+    val builder = new MetadataBuilder()
```

Comment:
maybe use TypeChange.toMetadata directly?

Reply:
The issue is that it could hide bugs in TypeChange.toMetadata. e.g. we wouldn't see if fromType gets serialized to INT instead of integer. That's why I pass all the args as plain strings and also explicitly write out the metadata keys instead of reusing the constants in TypeChangeMetadata: these should never change, if they do we want tests to fail.